trino/integration_test.go

```diff
@@ -1008,82 +1008,123 @@ func TestIntegrationUnsupportedHeader(t *testing.T) {
 	}
 }
 
-func TestIntegrationQueryContextCancellation(t *testing.T) {
-	err := RegisterCustomClient("uncompressed", &http.Client{Transport: &http.Transport{DisableCompression: true}})
-	if err != nil {
-		t.Fatal(err)
+func TestIntegrationQueryContext(t *testing.T) {
+	tests := []struct {
+		name           string
+		timeout        time.Duration
+		manualCancel   bool
+		expectedErrMsg string
+	}{
+		{
+			name:           "Context Cancellation",
+			timeout:        0,
+			manualCancel:   true,
+			expectedErrMsg: "canceled",
+		},
+		{
+			name:           "Context Deadline Exceeded",
+			timeout:        3 * time.Second,
+			manualCancel:   false,
+			expectedErrMsg: "context deadline exceeded",
+		},
 	}
-	dsn := *integrationServerFlag
-	dsn += "?catalog=tpch&schema=sf100&source=cancel-test&custom_client=uncompressed"
-	db := integrationOpen(t, dsn)
-	defer db.Close()
 
-	ctx, cancel := context.WithCancel(context.Background())
-	errCh := make(chan error, 3)
-	done := make(chan struct{})
-	longQuery := "SELECT COUNT(*) FROM lineitem"
-	go func() {
-		// query will complete in ~7s unless cancelled
-		rows, err := db.QueryContext(ctx, longQuery)
-		if err != nil {
-			errCh <- err
-			return
-		}
-		rows.Next()
-		if err = rows.Err(); err != nil {
-			errCh <- err
-			return
-		}
-		close(done)
-	}()
-
-	// poll system.runtime.queries and wait for query to start working
-	var queryID string
-	pollCtx, pollCancel := context.WithTimeout(context.Background(), 1*time.Second)
-	defer pollCancel()
-	for {
-		row := db.QueryRowContext(pollCtx, "SELECT query_id FROM system.runtime.queries WHERE state = 'RUNNING' AND source = 'cancel-test' AND query = ?", longQuery)
-		err := row.Scan(&queryID)
-		if err == nil {
-			break
-		}
-		if err != sql.ErrNoRows {
-			t.Fatal("failed to read query id", err)
-		}
-		if err = contextSleep(pollCtx, 100*time.Millisecond); err != nil {
-			t.Fatal("query did not start in 1 second")
-		}
-	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := RegisterCustomClient("uncompressed", &http.Client{Transport: &http.Transport{DisableCompression: true}})
```

Comment:
Pull all the code common to all tests before the loop. It will make the diff smaller, and make the test easier to read - the loop will only contain the test action and assertion, not the setup.